src/wasm/component/client/future.rs

```diff
@@ -0,0 +1,192 @@
+    task::{ready, Context, Poll},
+};
+
+use futures_core::Future;
+use wasi::http::{
+    outgoing_handler::{FutureIncomingResponse, OutgoingRequest},
+    types::{OutgoingBody, OutputStream},
+};
+
+use crate::{Body, Request, Response};
+
+/// A [`Future`] implementation for a [`Response`] that uses the [`wasi::io::poll`]
+/// primitives to poll receipt of the HTTP response.
+#[derive(Debug)]
+pub struct ResponseFuture {
+    request: Request,
+    state: RequestState,
+}
+
+impl ResponseFuture {
+    pub fn new(mut request: Request, outgoing_request: OutgoingRequest) -> crate::Result<Self> {
+        let state = match request.body_mut().take() {
+            Some(body) => {
+                let Ok(outgoing_body) = outgoing_request.body() else {
+                    return Err(crate::error::request("outgoing body error"));
+                };
+
+                let Ok(stream) = outgoing_body.write() else {
+                    return Err(crate::error::request("outgoing body write error"));
+                };
+
+                match wasi::http::outgoing_handler::handle(outgoing_request, None) {
+                    Ok(future) => RequestState::Write(RequestWriteState {
+                        response_future: Some(future),
+                        outgoing_body: Some(outgoing_body),
+                        stream: Some(stream),
+                        body,
+                        bytes_written: 0,
+                    }),
+                    Err(e) => return Err(crate::error::request("request error")),
+                }
+            }
+            None => match wasi::http::outgoing_handler::handle(outgoing_request, None) {
+                Ok(future) => RequestState::Response(future),
+                Err(e) => return Err(crate::error::request("request error")),
+            },
+        };
+
+        Ok(Self { request, state })
+    }
+}
+
+impl Future for ResponseFuture {
+    type Output = crate::Result<Response>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.get_mut();
+
+        match &mut this.state {
+            RequestState::Write(write_state) => match ready!(Pin::new(write_state).poll(cx)) {
+                Ok(future) => {
+                    this.state = RequestState::Response(future);
+                    Pin::new(this).poll(cx)
+                }
+                Err(e) => return Poll::Ready(Err(e)),
+            },
+            RequestState::Response(future) => {
+                if !future.subscribe().ready() {
+                    // NOTE(brooksmtownsend): We shouldn't be waking here since we don't know that
+                    // the future is ready to be polled again. Sleeping for a nanosecond appears to
+                    // allow the future to be polled again without causing a busy loop.
+                    std::thread::sleep(std::time::Duration::from_nanos(1));
+                    cx.waker().wake_by_ref();
+                    return Poll::Pending;
```

Comment:
This sleep is the only part of this PR that needs to be addressed before review. We should consider storing the waker & using a background task to wait for the future to be ready before waking the context. Not sleeping for a nanosecond here will infinitely hot-loop polling the future